### PR TITLE
Implementation of `Octo pr reviews`

### PR DESCRIPTION
**Pull request by pwntester**

### CODE REVIEW

File: lua/octo/menu.lua [hunk 1]
```diff
@@ -619,57 +619,193 @@ local changed_files_previewer =
 )
 
 local function changed_files()
-  local repo, number = util.get_repo_number()
+  local repo, number, _ = util.get_repo_number_pr()
```
Comment (pwntester) [inline lua/octo/menu.lua:622]: test comment

File: lua/octo/commands.lua [hunk 2]
```diff
@@ -60,6 +60,9 @@ local commands = {
     end,
     ready = function()
       M.pr_ready_for_review()
+    end,
```
Comment (pwntester) [inline lua/octo/commands.lua:63]: addition comment

File: lua/octo/fugitive.lua [hunk 3]
```diff
@@ -39,7 +39,6 @@ end
 
 function M.clean_fugitive_buffers()
   for _, w in ipairs(api.nvim_list_wins()) do
-    --if w ~= api.nvim_get_current_win() and vim.startswith(api.nvim_buf_get_name(api.nvim_win_get_buf(w)), "fugitive:") then
```
Comment (pwntester) [inline lua/octo/fugitive.lua:42]: deletion comment

File: lua/octo/menu.lua [hunk 4]
```diff
@@ -18,6 +18,8 @@ local json = {
   parse = vim.fn.json_decode
 }
 
+local M = {}
```
Comment (pwntester) [inline lua/octo/menu.lua:21]: use an export dict

Comment (pwntester) [reply lua/octo/menu.lua:21]: just a sample reply

File: lua/octo/menu.lua [hunk 5]
```diff
@@ -182,7 +184,7 @@ local function gen_from_issue()
   end
 end
 
-local function issues(repo, opts)
+function M.issues(repo, opts)
```
Comment (pwntester) [inline lua/octo/menu.lua:187]: another comment

**Review by pwntester — COMMENTED** [review]
fix stuff

**Review by PatOConnor43 — COMMENTED** [review]
This is looking good but I ran into an issue when trying to load my own PENDING review on this PR. I had this branch checked out and when I tried to select my review I got an error saying `Error executing vim.schedule lua callback: ...m/site/pack/packer/start/octo.nvim/lua/octo/fugitive.lua:207: Cursor position outside buffer`

Here's a picture of what it looked like:
image

I'm not exactly sure why those bubbles are at the top of vim. I'm talking about the 
`octo://pwntester/octo.nvim/56` and `lua/octo/menu.lua`. Are those vim tabs?

File: lua/octo/menu.lua [hunk 6]
```diff
@@ -618,58 +621,189 @@ local changed_files_previewer =
...
+          gh.run(
+            {
+              args = {"api", url},
+              cb = function(output, stderr)
+                if stderr and not util.is_blank(stderr) then
+                  api.nvim_err_writeln(stderr)
+                elseif output then
+                  local results = json.parse(output)
+                  for _, comment in ipairs(results) do
+                    api.nvim_buf_set_lines(self.state.bufnr, -1, -1, false, {"--"})
```
[the remark this answers is not in the document]
Comment (pwntester) [reply lua/octo/menu.lua:735]: yep, I was basically experimenting with the UI appareance but will use a single list and then call set_lines just one. thanks

**Comment by pwntester** [comment]
> Error executing vim.schedule lua callback: ...m/site/pack/packer/start/octo.nvim/lua/octo/fugitive.lua:207: Cursor position outside buffer

that is probably the call to `api.nvim_win_set_cursor` using the wrong win id or off by one row, will review it.

> I'm not exactly sure why those bubbles are at the top of vim. I'm talking about the
octo://pwntester/octo.nvim/56 and lua/octo/menu.lua. Are those vim tabs?

Im using a new tab to hold the review layout, so its probably the way your tabline shows tabs. Try creating a new tab with `tabnew` to reproduce

File: lua/octo/menu.lua (hunk 4 above)
Comment (pwntester) [reply lua/octo/menu.lua:21]: FOO

Comment (pwntester) [reply lua/octo/menu.lua:21]: FOO

Comment (pwntester) [reply lua/octo/menu.lua:21]: FOO

Comment (pwntester) [reply lua/octo/menu.lua:21]: FOO

Comment (pwntester) [reply lua/octo/menu.lua:21]: FOO

**Comment by pwntester** [comment]
test

File: lua/octo/menu.lua (hunk 4 above)
Comment (pwntester) [reply lua/octo/menu.lua:21]: test

Comment (pwntester) [reply lua/octo/menu.lua:21]: test

Comment (pwntester) [reply lua/octo/menu.lua:21]: test

Comment (pwntester) [reply lua/octo/menu.lua:21]: dfasdfasdf

Comment (pwntester) [reply lua/octo/menu.lua:21]: dasdfasdf

Comment (pwntester) [reply lua/octo/menu.lua:21]: asdf

**Comment by pwntester** [comment]
The `Octo pr reviews` command is ready for testing.

https://i.imgur.com/yEO5iUp.png
 
the window showing the comments and replies is similar to an issue buffer, you can use octo `comment` and `reaction` subcommands to add replies and reactions.
Also, you can use [c and ]c to jump between comments, [q and ]q for jumping to comments in a different file and <C-c> to clean up the review tab.

@lukas-reineke @PatOConnor43 @tami5 if you have some time, please play with it and let me know any bugs/feedback.

**Comment by tami5** [comment]
Hey, @pwntester `Octo review start` results in 
 
```viml
Error executing vim.schedule lua callback: ...vim/site/pack/packer/opt/octo.nvim/lua/octo/fugitive.lua:54: Vim(cc):E42: No Errors`
```

after opening a new tab.

**Comment by pwntester** [comment]
> Hey, @pwntester `Octo review start` results in after a new tab opened.
> 
> ```viml
> Error executing vim.schedule lua callback: .../octo/fugitive.lua:54: Vim(cc):E42: No Errors`
> ```

Yep, `Octo review start` is still WIP but that error seems weird. I will add a pcall to catch the error but, shouldnt all PRs have at least one change?
can you share the repo/pr you run `Octo review start` on to reproduce the problem? (if its public ofc)


> Edit: Okay `octo pr reivews` worked, but while choosing an item, I got the following error:
>
> ```viml
> Error executing vim.schedule lua callback: .../octo/reviews.lua:183: attempt to index a nil value
> ```

This is equivalent to the other error you reported, it seems like the qf window is empty, meaning that the review has no comments.
Can you share the repo/pr/review to reproduce the issue?

> Edit2: feedback
>
> - The qf menu should be 0.2 the size of the screen.
> - silent `]c/]q` or perhaps print small msg (e.g. next comment)
> - `]c/]q` should cycle, or perhaps provide an option to make it cycle if this isn't a desired default.
> - Couldn't comment on in the comments buffer, for some reason the mappings doesn't work.

> Can we visually select lines and comment on them yet.

For `Octo pr reviews` it makes no sense, but for `Octo review start` it is something that should be implemented for sure :)

> So far I'm amazed and amused by the experience, even now I've doing the reivew inside of vim. Amazing creation @pwntester, thanks a lot

Thanks! Yep, replying from nvim too! I find it really useful for my daily job where we use gh issues for ... everything :)

**Comment by tami5** [comment]
> can you share the repo/pr you run `Octo review start` on to reproduce the problem? (if its public ofc)

Sure just tried it here https://github.com/tami5/sql.nvim/pull/14

> Can you share the repo/pr/review to reproduce the issue?

I've actually tried it here by opening this pr and then choosing the first item from the list

> Yep, currently is hardcoded to 20 lines but should be configurable with a sane defalt like 0.2

hmm its a bit half screen on my end.

> The mappings are not enabled yet, but you should be able to do Octo comment add|delete and Octo reaction add hooray|...

I see, speaking of commands, fyi: there are no longer completion after checking out the pr.

> For `Octo pr reviews` it makes no sense, but for `Octo review start` it is something that should be implemented for sure :)

Nice!!

File: lua/octo/menu.lua (hunk 4 above)
Comment (pwntester) [reply lua/octo/menu.lua:22]: new comment

**Comment by pwntester** [comment]
@tami5 pushed some fixes to this branch:
- make qf height configurable and default to 0.2 vim height
- truncate qf entries
- add diff_hunk borders in review previews
- silent mappings
- map issue buffer mappings to review_comments buffer
- make ]c and ]q cycle

The only caveat is that I havent found the way to make ]c and [c cycle in the diff win (`Octo review start`) since this is a hardcoded mapping in vim, but I will look for a workaround.

Regarding the errors you were getting, I cannot reproduce them but I made some changes that could have fixed them. Can you check again?

File: after/syntax/octo_issue.vim [hunk 7]
```diff
@@ -12,7 +12,7 @@ syntax region issue_comment matchgroup=OctoNvimCommentDelimiters start=/\n\n/ ke
 
 hi def link OctoNvimDirty ErrorMsg
 hi def link OctoNvimCommentHeading PreProc
-hi def link OctoNvimCommentUser Underlined
```
Comment (pwntester) [inline after/syntax/octo_issue.vim:15]: comment 1

Comment (pwntester) [reply after/syntax/octo_issue.vim:15]: comment 2

File: lua/octo/menu.lua (hunk 4 above)
Comment (pwntester) [reply lua/octo/menu.lua:21]: test

Comment (pwntester) [reply lua/octo/menu.lua:21]: test

File: after/syntax/octo_issue.vim (hunk 7 above)
Comment (pwntester) [reply after/syntax/octo_issue.vim:15]: comment 3

**Comment by tami5** [comment]
I'll do some testing tom, thanks @pwntester .

**Comment by pwntester** [comment]
Thanks @tami5 

I changed a couple of things since you last tested it:
- there is no longer a review menu. `Octo pr reviews` will directly show the new tab with all the review threads in the qf window and select the first one.
- review threads that are Resolved and/or Outdated are now flagged as that in the QF window:
-

**Comment by tami5** [comment]
> there is no longer a review menu. Octo pr reviews will directly show the new tab with all the review threads in the qf window and select the first one.

Oh that's a wise decision 

> review threads that are Resolved and/or Outdated are now flagged as that in the QF window:

:heart_eyes:  :heart_eyes:

File: lua/octo/menu.lua (hunk 4 above)
Comment (pwntester) [reply lua/octo/menu.lua:21]: test

**Comment by tami5** [comment]
- I like how

File: lua/octo/menu.lua (hunk 4 above)
Comment (tami5) [reply lua/octo/menu.lua:21]: This awesome

**Comment by pwntester** [comment]
> - **REC**: the new menu works flawlessly. I really appreciate the highlight work 
 > - **REC**: diff_hunk borders in review previews are amazing.

 Thanks!

 > - **REC**: Better performance for some reason
 
 Moved to using the GraphQL API rather than the REST one, so I managed to reduce the number of calls to GH

 > - **ISSUE**: For some reason the highlight for this thread is broken. can u confirm?

Can you post a screenshot using the latest version of this PR?

 > - **ISSUE**: error with `[c` `]c`: Error executing lua ...nvim/site/pack/packer/opt/octo.nvim/lua/octo/reviews.lua:334: Key not found: pr_comments

I forgot to remove that code, it no longer makes sense since review threads are no longer grouped by reviews and therefore [q and ]q are enough.

 > - **ISSUE**: gh: Insufficient scopes for reacting to this Pull Request Review Comment. (HTTP 403), fyi it worked after checking github\

 This seems to be a problem with your personal token privs. Can you reproduce it?

 - **ISSUE**: Why spelling marking is disabled.
 - **FEAT**: I kind think now that the menu was good idea, but I'd add that having those awesome categorizes in them for filtering would be awesome, as well as fetching them all and putting in temp dict instead of individually loading the preview
 - **FEAT**: highlight the line and range corresponding to the diff, like tree-sitter highlight.  
 - **FEAT**: cn/cp should act like `[q]q`

**Pull request merged**